Propose a unicast method of configuring keepalived for Baremetal IPI (#230)
openshift/enhancements: enhancements/baremetal/baremetal-keepalived-unicast.md

---
title: Baremetal-Keepalived-Unicast
authors:
  - "@bcrochet"
reviewers:
  - TBD
approvers:
  - TBD
creation-date: 2020-02-21
last-updated: 2020-02-21
status: implementable
---

# baremetal-keepalived-unicast

## Release Signoff Checklist

- [x] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

The Baremetal platform uses keepalived to move VIPs between the bootstrap host and the
masters, then between the masters and the workers. This worked via multicast, but not
all environments have multicast available. This proposal seeks to rememdy that.

## Motivation

Currently, the Baremetal platform utilizes keepalived to maintain the VIPs necessary
for deployment and operation. It presently does so via multicast. This requires a
minimal amount of configuration, as the multicast peers do not need to be specified.
However, there are likely to be environments where multicast has been disabled because

> of network policy. In these cases, it is necessary to utilized unicast addresses within

Reviewer: am I right to assume that multicast also require a L2 network connectivity, which in turn means you can't:

1.  have nodes on different vlans (e.g. distribute your masters across failure domains)
2.  run multiple clusters on the same vlan

Reviewer 2: The multicast advertisements already include the cluster name so it's safe to run multiple clusters on the same network.  Any advertisements coming from other clusters would be ignored.

It's possible to set up multicast forwarding between different networks.  I have not seen anyone document an example for how to do so, though.

Reviewer 3: Even if the forwarding is possible, it would be another requirement on the customer, isn't it?

Reviewer 2: Yes.  I wasn't trying to imply that it's great, just answering ohad's question about what is possible

  address of the bootstrap host. This server will be run on the bootstrap host as
  a sidecar on the keepalived pod.

* The bootstrap keepalived pod previously used a (mostly) static configuration
  template. The keepalived monitor needs to run on the bootstrap host now. The
  same monitor will be used for both the bootstrap and the control plane.

### Risks and Mitigations

It is important to not destabilize our existing installation method, that uses
multicast successfully. Full parity is expected before landing.

## Design Details

### Test Plan

Because of the fundamental inclusion of this in Baremetal IPI installations,
any breakage by CI before landing can probably be assumed to be caused by
this work.

## Implementation History

A PR with implementation in baremetal-runtimecfg can be found here:

* https://github.com/openshift/baremetal-runtimecfg/pull/47

The Machine Config operator portion can be found here:

* https://github.com/openshift/machine-config-operator/pull/1504

## Drawbacks

* Multicast is a simpler configuration with fewer moving parts. However, not every
  environment allows multicast, whereas all environments would allow unicast.

> * The customer must keep the Etcd SRV record up to date in case they add/replace

Reviewer 3: I think there is a plan to remove the dependency on etcd SRV records

Author: That's good to know. Do you have a pointer to a discussion on that?

Reviewer 3: @crawford @hexfusion

Reviewer 4: Yes, we are dropping this dep in 4.4. etcd no longer is dependent on SRV in 4.4+

---
title: Baremetal-Keepalived-Unicast
authors:
  - "@bcrochet"
  - "@yboaron"
reviewers:
  - TBD
approvers:
  - TBD
creation-date: 2020-02-21
last-updated: 2020-02-21
status: implementable
---

> # baremetal-keepalived-unicast

Reviewer 2: There's a section in the template about upgrade/downgrade impact, as well as version skew handling.  I was expecting to see those topics explored in this doc.  Can you take a look at adding those?

Reviewer 5: Sure, good point. 
I'll add that to the PR.

Reviewer 5: Done

## Release Signoff Checklist

- [x] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

The Baremetal platform uses VIPs (Virtual IP) and [Keepalived](https://www.keepalived.org/) to 
support high availability for API and default Ingress.
A VIP is used to provide failover of the service across the relevant nodes (including the bootstrap instance).

Two VIPs are supported:
* api-vip
* ingress-vip

The `api-vip` first resides on the bootstrap instance and it moves to one of the control plane nodes, but only after the bootstrap process has completed and the bootstrap instance is stopped. This ensures that the API on the control plane nodes is fully functional before the api-vip moves.
The `ingress-vip` resides on one of the nodes that runs a router pod.

> In current implementation, Keepalived is configured to run [VRRP management protocol](https://tools.ietf.org/html/rfc3768#section-3) of the VIP addresses in a multicast mode, but since not all environments have multicast available, this proposal seeks to remedy that.

Reviewer 5: Done

> 

Reviewer 2: A helpful link to add here would be https://github.com/openshift/installer/blob/master/docs/design/baremetal/networking-infrastructure.md

That doc covers the overview you provide here in some more detail

Reviewer 5: Done

## Motivation

As mentioned above, currently Keepalived sends the VRRP management protocol via multicast. This requires a
minimal amount of configuration, as the multicast peers do not need to be specified.
However, there are likely to be environments where multicast has been disabled because
of network policy. 

Additionally, moving to unicast will also solve the possible collisions when two or more clusters are on the same broadcast/multicast domain. this use case described in more detail in `user story #2` below.

> In these cases, it is necessary to utilized unicast addresses within keepalived rather than make this an optional configuration. It makes sense to make it the default since unicast will "just work" in all environments.

Reviewer 2: ```suggestion
In these cases, it is necessary to utilize unicast addresses within keepalived rather than make this an optional configuration. It makes sense to make it the default since unicast will "just work" in all environments.
```

### Goals

* A deployment is possible without the need for multicast on a network.
* No loss of functionality from the current multicast implementation.

### Non-Goals

* A switchable implementation. It does not need to be an installer choice.
* The control plane hosts on different subnets (not on the same L2 segment).

## Proposal

To enable keepalived to communicate via unicast instead of multicast, and to handle
the fact that the bootstrap IP is unknown to the masters.

### User Stories

#### Story 1 - Allow deployment where networking doesn't permit multicast

As an administrator, I want to deploy OpenShift on bare metal nodes with 

> installer-provisioned infrastructure (IPI) where the networking doesn't permit multicast.

Reviewer 6: AFAICT this applies to all on-prem platforms regardless of IPI/UPI.

Reviewer 5: Do we want to adopt unicast change also to other on-prem platforms? 
I thought that there is no such requirement.

Reviewer 7: > AFAICT this applies to all on-prem platforms regardless of IPI/UPI.

I think UPI already implies that the users might be using their own networking setup not prescribed by OpenShift, like None platform or vSphere upi etc. So I think calling out baremetal-ipi (or others like that) seems better to me..

Reviewer 2: agree that clarifying if this will be bare metal ipi only, or also openstack IPI and vsphere IPI, would be good

Reviewer 5: I checked this issue with the PMs and seems that unicast requirement is relevant only for the baremetal-ipi case.

I updated the summary and goals sections.

Reviewer 5: done

Some customers have restrictions on multicast network traffic (it impacts telco/edge customers since it’s not possible to use multicast between the regional and far-edge locations).
The current deployment supports high availability for API and default ingress using multicast traffic for VRRP management of the VIP addresses.

#### Story 2 - Allow multiple deployments on the same network

For PoC and test deployments we might need to run multiple clusters on the same network (requiring a dedicated network is a heavy dependency in non-virtualized environments), while we’d still recommend isolated networks for production it would be desirable to enable deployment of multiple clusters on a single broadcast domain.
Keepalived uses the `virtual_router_id` field to specify to which VRRP router the [instance](https://www.keepalived.org/doc/configuration_synopsis.html#vrrp-instance-definitions-synopsis) belongs,`virtual_router_id` range is 1 - 255.

> Each baremetal deployment includes two VRRP routers:

Reviewer 6: ```suggestion
Each baremetal deployment includes two virtual routers:
```

* API VRRP router
* Ingress VRRP router

> Although the `virtual_router_id` values for both API and Ingress are calculated using [checksum function based on cluster name](https://github.com/openshift/baremetal-runtimecfg/blob/master/pkg/config/node.go#L162-#L171) we can still end up in a situation where two VRRP routers from different deployments share the same `virtual_router_id`.

Reviewer 6: ```suggestion
Although the `virtual_router_id` values for both API and Ingress are calculated using [checksum function based on cluster name](https://github.com/openshift/baremetal-runtimecfg/blob/master/pkg/config/node.go#L162-#L171) we can still end up in a situation where virtual routers from different deployments or services share the same `virtual_router_id`.
```

### Implementation Details/Notes/Constraints

> Unlike multicast mode, in unicast we should specify all the group members in 'unicast_peer' section at [Keepalived conf](https://www.keepalived.org/manpage.html) for each VIP. Lets first present briefly the supported VIPs and the group members for each one of them.

Reviewer 6: specify AND keep up to date

1. Update the keepalived conf template files for both [control plane nodes](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/templates/master/00-master/baremetal/files/baremetal-keepalived-keepalived.yaml#L49-#L73) and [compute nodes](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/templates/worker/00-worker/baremetal/files/baremetal-keepalived-keepalived.yaml) to include the unicast details.
2. The keepalived-monitor container should be updated to [retrieve](https://github.com/openshift/baremetal-runtimecfg/blob/8242fd84d6c6d267d2d5551abf8e4f778b4f0615/pkg/config/node.go#L219-#L244) the nodes (both compute and control) IP addresses from the kube-apiserver and render them to keepalived config.

### Risks and Mitigations

It is important to not destabilize our existing installation method, that uses
multicast successfully. Full parity is expected before landing.

## Design Details

### Test Plan

Because of the fundamental inclusion of this in Baremetal IPI installations,
any breakage by CI before landing can probably be assumed to be caused by

> this work.

Reviewer 6: I would include here tests that:
* Take down nodes both in active as in passive mode and observe the VIP move correctly and without significantly increasing the current expected downtime in VIP changes.
* Force a disaster recovery scenario and see that we can recover the cluster following the usual procedure.

Reviewer 5: Done

## Implementation History

A PR with implementation in baremetal-runtimecfg can be found here:

* https://github.com/openshift/baremetal-runtimecfg/pull/65

The Machine Config operator portion can be found here:

* https://github.com/openshift/machine-config-operator/pull/1768

## Drawbacks

* Multicast is a simpler configuration with fewer moving parts. However, not every
  environment allows multicast, whereas all environments would allow unicast.

## Alternatives

An alternative would be to absolutely require multicast for installation and

> operation.

Reviewer 6: We could also have a sort of "stretch goal" where the runtimecfg can check for a CRD in order to switch to VRRP mode for those environments that want to use it after installation/clustering.

Reviewer 5: Well, that means we have two different ways to control keepalived mode, which might lead to conflicts. I would prefer to have a single way to control Unicast/Multicast.

Is it possible to update the Keepalived mode by patching the MCO manifests?

---
title: Baremetal-Keepalived-Unicast
authors:
  - "@bcrochet"
  - "@yboaron"
reviewers:
  - TBD
approvers:
  - TBD
creation-date: 2020-02-21
last-updated: 2020-02-21
status: implementable
---

# baremetal-keepalived-unicast

## Release Signoff Checklist

- [x] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

Non cloud platforms use VIPs (Virtual IP) and [Keepalived](https://www.keepalived.org/) to support high availability for API and default Ingress.
For more information in the matter of VIPs please use the following [link](https://github.com/openshift/enhancements/blob/master/enhancements/network/baremetal-networking.md).

> In current implementation, Keepalived is configured to run [VRRP management protocol](https://tools.ietf.org/html/rfc3768#section-3) of the VIP addresses in a multicast mode, but since not all environments have multicast available, this proposal seeks to remedy that.

Reviewer 7: what i would like the summary to include is whether we will be replacing with new impl across the board or are we providing an alternate way for users to choose.

Reviewer 5: Done

## Motivation

As mentioned above, currently Keepalived sends the VRRP management protocol via multicast. This requires a
minimal amount of configuration, as the multicast peers do not need to be specified.
However, there are likely to be environments where multicast has been disabled because
of network policy. 

Additionally, moving to unicast will also solve the possible collisions when two or more clusters are on the same broadcast/multicast domain. This use case described in more detail in `user story #2` below.

In environments where there are lots of existing virtual routers managing Virtual IPs with multicast messaging or disallow the multicast VRRP traffic completely, it is necessary to configure Keepalived unicast communication  rather than making it an optional configuration. Making it default ensures a successful deployment, so It makes sense to make it the default and if we want to continue to support multicast VRRP, have it as an option.

### Goals

* Allow deployment in networks where multicast VRRP traffic is forbidden or non-functional.
* No loss of functionality from the current multicast implementation.

> * Prevent VRRP ID collisions during deployment.

Reviewer 7: if this is supposed to be changes for all clusters, we need to include a goal that can move all existing clusters to this new model without downtime...

Reviewer 5: Yep, I'll add that goal.
But I'm not sure we can promise zero downtime, even for planned failover we still have the time it takes for the new leader to publish himself.

Reviewer 5: Done

Each baremetal deployment includes two virtual routers:
* API
* Ingress

Although the `virtual_router_id` values for both API and Ingress are calculated using [checksum function based on cluster name](https://github.com/openshift/baremetal-runtimecfg/blob/master/pkg/config/node.go#L162-#L171) we can still end up in a situation where virtual routers from different deployments or services share the same `virtual_router_id`.

### Implementation Details/Notes/Constraints

Unlike multicast mode, in unicast we should specify and keep up to date all the group members in 'unicast_peer' section at [Keepalived conf](https://www.keepalived.org/manpage.html) for each VIP. Lets first present briefly the supported VIPs and the group members for each one of them.

As mentioned above, `api-vip` and `ingress-vip` are supported to enable high availability.

The `api-vip` should resides on the bootstrap instance during the bootstrap phase and move to one of the control plane nodes after the bootstrap process has completed and the bootstrap instance is stopped.

> The group members for the `api-vip` should include the bootstrap node and the control plane nodes.

Reviewer 7: > and the bootstrap instance is stopped.

hmm.. this seems odd, why would stopping the bootstrap host be required, why can we move as soon as the control-plane nodes have capacity to host the VIP?

Reviewer 5: My bad, the VIP moves to one of the control plane nodes as soon as he can host it.
I'll update that .

The bootstrap Keepalived pod includes:
- [Keepalived container](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.yaml#L35-#L93)
- [TCP server container](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.yaml#L27#L34) that will be responsible to return the bootstrap IP address.
- [keepalived-monitor container](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.yaml#L94-#L138) is responsible to render the Keepalived config file.

The keepalived-monitor container in bootstrap node will first render Keepalived cfg file with an empty 'unicast_peer' section, as a result of that the `api-vip` will be set to the bootstrap node by Keepalived.

At the next step, the bootstrap's keepalived-monitor container will repeatedly retrieve the control plane nodes details from the localhost:kube-apiserver and add them to 'unicast_peer' section.

> The bootstrap node owns the `api-vip' during bootstrap phase because it's configured with [higher](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.conf.tmpl#L13) VRRP priority than the control plane nodes [priority](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/templates/master/00-master/baremetal/files/baremetal-keepalived-keepalived.yaml#L26).

Reviewer 7: ```suggestion
The bootstrap node owns the `api-vip` during bootstrap phase because it's configured with [higher](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.conf.tmpl#L13) VRRP priority than the control plane nodes [priority](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/templates/master/00-master/baremetal/files/baremetal-keepalived-keepalived.yaml#L26).
```

Reviewer 5: Done

##### Control plane nodes

The [Keepalived pod](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/templates/common/baremetal/files/baremetal-keepalived.yaml) in control plane nodes already includes Keepalived and keepalived-monitor containers. 
To support unicast mode, the keepalived-monitor should be updated to include the [bootstrap node IP address](https://github.com/openshift/baremetal-runtimecfg/blob/8242fd84d6c6d267d2d5551abf8e4f778b4f0615/pkg/monitor/dynkeepalived.go#L30) and [IP addresses of control plane nodes](https://github.com/openshift/baremetal-runtimecfg/blob/8242fd84d6c6d267d2d5551abf8e4f778b4f0615/pkg/config/node.go#L336-#L380).

> 

Reviewer 7: this should include information that the bootstrap ip is being fetching by `GET <apiVIP>:<ip discovery port>`

Is this discovery service running on all the control-plane nodes too, because in case i want to add a new control-plane machine, it will try to reach out to this IP:PORT for an ip and will probably fail to get all the information to join the cluster...

Reviewer 5: That should be OK, the control-pane node repeatedly tries to GET bootstrap IP from <api-vip><discovery-port> in case no reply was received (after a timeout), Keepalived will be configured in this node only with the control-plane nodes as peers.

Reviewer 5: Done

The current deployment supports high availability for API and default ingress using multicast traffic for VRRP management of the VIP addresses.

#### Story 2 - Allow multiple deployments on the same network

For PoC and test deployments we might need to run multiple clusters on the same network (requiring a dedicated network is a heavy dependency in non-virtualized environments), while we’d still recommend isolated networks for production it would be desirable to enable deployment of multiple clusters on a single broadcast domain.
Keepalived uses the `virtual_router_id` field to specify to which VRRP router the [instance](https://www.keepalived.org/doc/configuration_synopsis.html#vrrp-instance-definitions-synopsis) belongs,`virtual_router_id` range is 1 - 255.
Each baremetal deployment includes two virtual routers:
* API
* Ingress

Although the `virtual_router_id` values for both API and Ingress are calculated using [checksum function based on cluster name](https://github.com/openshift/baremetal-runtimecfg/blob/master/pkg/config/node.go#L162-#L171) we can still end up in a situation where virtual routers from different deployments or services share the same `virtual_router_id`.

> ### Implementation Details/Notes/Constraints

Reviewer 7: will it be useful to describe some recovery scenarios?

for example,
let's say we lose 2 out of 3 etcd, here we could loose access to kube-apiserver for getting information, how would the recovery process look like?

Reviewer 5: I'm not familiar with recovery scenarios.
I'll clarify this point and in case it's relevant for Keepalived mode I'll add it to this enhancement.

#### ingress-vip implementation

Add unicast support for the `ingress-vip` is simpler than the `api-vip`.

The following steps should be done to enable unicast for `ingress-vip`:
1. Update the Keepalived conf template files for both [control plane nodes](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/templates/master/00-master/baremetal/files/baremetal-keepalived-keepalived.yaml#L49-#L73) and [compute nodes](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/templates/worker/00-worker/baremetal/files/baremetal-keepalived-keepalived.yaml) to include the unicast details.
2. The keepalived-monitor container should be updated to [retrieve](https://github.com/openshift/baremetal-runtimecfg/blob/8242fd84d6c6d267d2d5551abf8e4f778b4f0615/pkg/config/node.go#L219-#L244) the nodes (both compute and control) IP addresses from the kube-apiserver and render them to Keepalived config.

### Risks and Mitigations

It is important to not destabilize our existing installation method, that uses
multicast successfully. Full parity is expected before landing.

## Design Details

> ### Test Plan

Reviewer 2: A risk here is that we don't have upgrade coverage for bare metal IPI in CI.  We may want to make that a prerequisite to landing this work.

Reviewer 2: can you add a note on this?

Reviewer 5: I already added a note in the risks section regarding the upgrade, see [1].

Please let me know if you think the note should be updated.

[1] https://github.com/openshift/enhancements/blob/fb8a5aa76c8db40f53d2b363a6d9fd1ff2556c40/enhancements/baremetal/baremetal-keepalived-unicast.md#risks-and-mitigations

any breakage by CI before landing can probably be assumed to be caused by
this work.

Take down nodes both in active as in passive mode and observe the VIP move correctly and without significantly increasing the current expected downtime in VIP changes.

Force a disaster recovery scenario and see that we can recover the cluster following the usual procedure.

Since the baremetal-runtimecfg is used by all on-prem platforms, we should verify that Ovirt, vSphere, and OpenStack still successfully deployed in Keepalived multicast mode.

### Upgrade / Downgrade Strategy

In case that the Keepalived config file was already generated by the previous SW version, Keepalived will first continue to run in a mode determined in accordance with the existing file.
If this mode is different from the mode specified in MCO templates a Keepalived configuration migration process should be triggered.

The next steps illustrate Keepalived flow in case of upgrade/downgrade:
1. If config file exists determined_mode=mode determined from the file else determined_mode=mode specified in MCO templates
2. Run Keepalived in determined_mode.

> 3. If determined_mode != desired_mode (mode specified in MCO templates) initiate a configuratiom migration process.

Reviewer 2: ```suggestion
3. If determined_mode != desired_mode (mode specified in MCO templates) initiate a configuration migration process.
```

#### Configuration migration process

Since multicast and unicast are considered as separate Keepalived domains and will not interoperate, the updated configuration should simultaneously reload in all the relevant Keepalived instances.

> [TBD] How to implement simultaneously reload of multiple Kepealived instances.

Reviewer 2: This seems like a pretty critical design point.  I don't think this enhancement is actually "implementable" with the upgrade process still TBD

Reviewer 5: I added a description of the suggested sync reload steps.

As per the `ingress-vip`, it should resides on a node (either control or compute) that runs a router pod, so for that case the group members should include all control and compute nodes.

#### api-vip implementation

Since the bootstrap IP address is not known until the server is up, a method for 
retrieving that node will need to be implemented to allow the control plane nodes to include the 
bootstrap IP address in their Keepalived configuration. 

##### Bootstrap node

The bootstrap [Keepalived](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.yaml) instance runs as static pod and the relevant [Keepalived configuration](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.conf.tmpl) rendered by the [baremetal-runtimecfg](https://github.com/openshift/baremetal-runtimecfg).

The bootstrap Keepalived pod includes:
- [Keepalived container](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.yaml#L35-#L93)

> - [TCP server container](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.yaml#L27#L34) that will be responsible to return the bootstrap IP address.

Reviewer 2: This is probably outdated?

Reviewer 5: When running keepalived (for API-VIP) in unicast mode during bootstrap phase we need to set the following : 

*  In bootstrap node, add masters nodes IPs to 'unicast_peer' in keepalived conf 
*  In master nodes, add other masters IPs + bootstrap node IP address to  'unicast_peer' in keepalived conf.

We retrieve the master nodes details from OCP-API, and we need to find a way to retrieve somehow the bootstrap IP address, so, we decided to run a TCP server only on the bootstrap node for that purpose.

Do you know if the bootstrap IP address could be retrieved in some other way?

Reviewer 2: The bootstrap node itself could provide it when generating manifests?

Reviewer 5: @celebdor  Do you think ^^ is feasible?

Reviewer 6: I don't know. Did you check the MCO bootstrap and the installer bootstrap to check it?

Reviewer 5: I checked with MCO and installer teams, seems that the bootstrap IP could be retrieved without running a dedicated server in the bootstrap node for that purpose. 
One option is to read it from etcd-endpoints ConfigMap, I updated the enhancement (the code should be updated).

Reviewer 5: Done

Since the baremetal-runtimecfg is used by all on-prem platforms, we should verify that Ovirt, vSphere, and OpenStack still successfully deployed in Keepalived multicast mode.

### Upgrade / Downgrade Strategy

In case that the Keepalived config file was already generated by the previous SW version, Keepalived will continue to run in a mode determined in accordance with the existing file.
If the cluster admin wants to change the Keepalived mode post upgrade process, he should trigger a configuration migration process.

The next steps illustrate Keepalived flow in case of upgrade/downgrade:
1. If config file exists determined_mode=mode determined from the file else determined_mode=mode specified in MCO templates
2. Run Keepalived in determined_mode.

#### Configuration migration process

Since multicast and unicast are considered as separate Keepalived domains and will not interoperate, the updated configuration should simultaneously reload in all the relevant Keepalived instances. This option is feasible since the nodes run NTP.

The next steps describe how the Keepalived mode should be changed in the cluster:

> 1. Cluster admin should create a MachineConfig CR, this CR will generate a new file in all nodes.

Reviewer 2: I think we should have this be an automated process that no admin needs to deal with.  Any thoughts on how that might be done?

Reviewer 5: Yep, we can use MCO node's annotation as @celebdor suggested.
I'll update this section

Reviewer 5: Done

If this mode is different from the mode specified in MCO templates a Keepalived configuration migration process should be triggered.

The next steps illustrate Keepalived flow in case of upgrade/downgrade:
1. If config file exists determined_mode=mode determined from the file else determined_mode=mode specified in MCO templates
2. Run Keepalived in determined_mode.
3. If determined_mode != desired_mode (mode specified in MCO templates) initiate a configuration migration process.

#### Configuration migration process

Since multicast and unicast are considered as separate Keepalived domains and will not interoperate, the updated configuration should simultaneously reload in all the relevant Keepalived instances. This option is feasible since the nodes run NTP.

The next steps describe how synced reload should be implemented:

1. One of the nodes will take the role of coordinator (the node with the lowest IP address).
2. The coordinator should be responsible for:
    - Deciding if simultaneously reload is required.
    - Verifying that all nodes are ready for reload.

>     - Choosing the timing for reload and updating the rest of the nodes.

Reviewer 2: How will updating the rest of the nodes be done?

Reviewer 5: A new TCP server will be added to the keepalived-monitor container, this server will serve as a kind of control channel.
The coordinator node will use this server for communicating with rest of the nodes.

Reviewer 2: I'm not too comfortable with the idea of a new custom bespoke TCP service just for this use case.  We should always hesitate before adding new network services.

Have you considered a more kubernetes native approach to this problem?  I imagine pods all looking at a shared configmap that indicates what mode and configuration they should be running in, something like that?

In other words, what alternatives have been considered here?

Reviewer 5: @russellb 

Well, currently we don't use any CRD nor configmap for configuring and running keepalived (Keepalived should run early in bootstrap phase to provide access to temp control plane), we use static pod capability and configuration is rendered by the MCO. 

An alternative method could be to use the MachineConfig CRD for triggering the migration, 
create MachineConfig CR, this CR will create a new file (includes the desired mode + planned reload time) in all nodes and keepalived-monitor will do the rest of the work.

What do you think about this option?

Reviewer 2: Yes, that sounds better.  I'd just like to see something built around existing interfaces

Reviewer 6: The original idea was to just have a MachineConfig that writes a config file that the keepalived monitor reads. Once the monitor reads it and notices that the requested mode differs from the configured one, it checks if the other nodes are in the same Node object `CurrentConfig`. Once all the monitors see that the other nodes are in the same currentConfig, they should be able to perform the reload.

This solution just uses the OCP API, i.e., a MachineConfig and the Node annotation for MC rendered currentconfig and desiredconfig.

This means that the user would create a MachineConfig and once all the machines have rebooted and have the new config, the change can take place. Personally I'd prefer if it was optional to set a time for when to do the reload, as it may involve a bit of guesswork of knowing when the MachineConfig rollout will end (to know the earliest time).
